### DISCUSSION TITLE
[App Router] Where is the best place to validate route's dynamic params?

### DISCUSSION THREAD

**Discussion by amirhhashemi** [vercel/next.js#52602] [Help] [answered]
### Summary

I want to validate route's dynamic params and return an 404 error if it's not valid. For example for `/users/[userId]` I want to return 404 if `userId` is not exactly 12 characters. Where should I do that? In `page.tsx`, `layout.tsx` or `template.tsx`?

Personally I think `template.tsx` is the best option because I have multiple pages that share the same dynamic segment and I can define a single `template.tsx` file that is shared between all of them. Also it re-runs on navigation so I can re-validate the params after navigation which is not possible with `layout.tsx`. (navigating from `/users/1` to `/users/12` doesn't re-render `/users/layout.tsx`, right?)

But I'm a little espectical. Consider these pages.
- `/users/[userId]/page.tsx`
- `/users/[userId]/posts/[postId]/page.tsx`

If I want to use `template.tsx` I have to validate the params in two places: 
- `/users/[userId]/template.tsx`: validate `userId`
- `/users/[userId]/posts/[postId]/template.tsx`: validate `userId` and `postId`

Which means if the user navigates to `/users/1/posts/2`, `userId` is validated twice. It's not a big deal at all; it's almost nothing but I don't like the duplication in the validation logic.

I know it's not necessary to think about these things. There are way more important things to worry about. But I'm new to the app router and I though maybe I learn something new, so that's why I'm asking it here.

A side question: how to access route params in server components? All the hooks that provide route params are client only! (or i'm blind:) )

**Comment by robertpiosik** [comment, marked as answer]
Templates are just special layouts that recreates upon navigation, they have nothing special that can help with the case.

What you want to do is utilise "route groups", this way you will split the flow of rendering so layout (or template if you need that rerender) that holds validation for `/users/[userId]` won't run when rendering `/users/[userId]/posts/[postId]`.

`app/users/[userId]/(user)/layout.tsx` <- layout for a user if you need one (ui, providing redux etc.)
`app/users/[userId]/(user)/(user-id-validation)/layout.tsx`
`app/users/[userId]/(user)/(user-id-validation)/page.tsx`
`app/users/[userId]/(user)/posts/[postId]/(user-id-and-post-id-validation)/layout.tsx`
`app/users/[userId]/(user)/posts/[postId]/(user-id-and-post-id-validation)/page.tsx`

This split is actually a good thing because in the post route you can fetch data differently. By using [parallel data fetching](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching#data-fetching-patterns) you will serve data way faster.

**Comment by rasif-sahl** [comment]
To validate route dynamic parameters and return a 404 error if they are not valid, you can utilize Next.js's built-in server-side routing capabilities and custom error handling.

Here's an approach you can follow:

1. Create a custom error page (404.js) in the pages directory. This page will be displayed when a 404 error occurs.
```
// pages/404.js

function NotFoundPage() {
  return <h1>404 - Page Not Found</h1>;
}

export default NotFoundPage;
```
2. In your dynamic route pages (e.g., pages/users/[userId]/page.js, pages/users/[userId]/posts/[postId]/page.js), validate the dynamic route parameters within the getServerSideProps function.

```
// pages/users/[userId]/page.js

export async function getServerSideProps({ params }) {
  const { userId } = params;

  // Validate the userId
  if (userId.length !== 12) {
    return {
      notFound: true, // Return 404 error
    };
  }

  // Continue with normal server-side rendering logic
  // ...

  return {
    props: {
      // Provide data for rendering the page
    },
  };
}

function UserPage({ /* data */ }) {
  // Render the page content
}

export default UserPage;
```
In this example, the getServerSideProps function is used to validate the userId parameter. If the validation fails (e.g., the length is not exactly 12), the notFound property is set to true, triggering the display of the custom 404 page.

By performing the validation within getServerSideProps, you can ensure that the route parameters are validated on the server side before rendering the page.

Regarding your side question, accessing route parameters in server components can be done using the useRouter hook from 'next/router'. However, this hook is only available on the client side, not on the server side.

If you need to access route parameters within server components, you can pass the parameters as props or retrieve them from the request object (req) in server-side APIs.

Note that the getServerSideProps function runs on the server side during the initial rendering, while the useRouter hook and other client-side hooks are available during subsequent client-side navigation.

By following this approach, you can validate route parameters and return a 404 error when they are not valid, while also avoiding duplication of validation logic across different dynamic route pages.

**Reply by robertpiosik** [reply]
Is this a chatbot response? 🤓 OP asked specifically about the App Router.
